### PR TITLE
[None][refactor] refactor the CUDA graph runner to manage all CUDA graphs

### PR DESCRIPTION
**Pull request by QiJune**
## Summary by CodeRabbit

* **Refactor**
  * Centralized CUDA graph lifecycle and memory management, improving performance, stability, and cleanup.
  * Faster warmup and reduced memory fragmentation through pooled allocations and unified release.
  * Streamlined forward path with reliable fallback to eager execution when graphs aren’t applicable.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by amukkara — APPROVED** [review]
Approving for phi3 changes.

**Review by leslie-fang25 — APPROVED** [review]

**Review by lfr-0531 — COMMENTED** [review]
A few minor issues, everything else is LGTM!

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 1]
```diff
@@ -25,118 +35,288 @@ def is_graph_capturing() -> bool:
...
+    def needs_capture(self, batch_size: int):
+        return (batch_size, self.draft_len) not in self.graph_outputs
+
+    def capture_graph(self, batch_size: int, forward_fn: Callable,
+                      initial_inputs: Dict[str, Any]):
+        """Captures the forward pass for a given batch size."""
+        engine = self._get_engine()
+        key = (batch_size, self.draft_len)
+
+        max_tokens_per_req = 1
```
[the remark this answers is not in the document]
Comment (QiJune) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:145]: Removed

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 2]
```diff
@@ -539,10 +538,12 @@ def wrapper(self, *args, **kwargs):
     def no_cuda_graph(self):
         _run_cuda_graphs = self._run_cuda_graphs
         self._run_cuda_graphs = False
```
[the remark this answers is not in the document]
Comment (QiJune) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Yes, removed

File: tests/unittest/_torch/helpers.py [hunk 3]
```diff
@@ -162,3 +163,96 @@ def block_scale_gemm(mat_a: torch.Tensor, mat_scale_a: torch.Tensor,
...
+
+def set_graph_capturing(enable: bool):
+    _local.is_graph_capturing = enable
+
+
+def is_graph_capturing() -> bool:
+    return _local.is_graph_capturing
+
+
+class DecodingCUDAGraphRunner:
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tests/unittest/_torch/helpers.py:185]: We don't need a full model engine for testing. Since only the `capture/replay` APIs of `CUDAGraphRunner` is used in these tests, we only need to create a mock engine with `use_mrope/spec_config/is_spec_decode/pytorch_backend_config/_cuda_graph_batch_sizes/_max_cuda_graph_batch_size/max_beam_width/_cuda_graph_mem_pool` fields, all other fields/methods are unnecessary. All of these fields have simple default values in single-batch test case.

[the remark this answers is not in the document]
Comment (mikeiovine) [reply tests/unittest/_torch/helpers.py:185]: +1 for deleting the legacy code. Another alternative to mocking `ModelEngine` is to make it so that we can easily create `ModelEngine`s from arbitrary `torch.nn.Module`s. Instead of requiring an HF config, you can add the option to take a module with the right forward signature.

**Review by mikeiovine — COMMENTED** [review]
It looks correct to me, but what's the motivation for the refactor? It's good that all of `ModelEngine`'s CUDA graph related code is now in one centralized place for sure, but I'm a little leery about coupling `CUDAGraphRunner` and `ModelEngine` so tightly. I mean that we're in a circular dependency situation now: `ModelEngine` depends on and uses `CUDAGraphRunner` and `CUDAGraphRunner` depends on and uses `ModelEngine`. This can make the logic a bit hard to follow because the control flow now alternates between `ModelEngine` and the `CUDAGraphRunner` in many places.

**Comment by QiJune** [comment]
> It looks correct to me, but what's the motivation for the refactor? It's good that all of `ModelEngine`'s CUDA graph related code is now in one centralized place for sure, but I'm a little leery about coupling `CUDAGraphRunner` and `ModelEngine` so tightly. I mean that we're in a circular dependency situation now: `ModelEngine` depends on and uses `CUDAGraphRunner` and `CUDAGraphRunner` depends on and uses `ModelEngine`. This can make the logic a bit hard to follow because the control flow now alternates between `ModelEngine` and the `CUDAGraphRunner` in many places.

Hi @mikeiovine , the motivation is to make the `CUDAGraphRunner` manage all the cuda graphs, thus,`ModelEngine` does not need to know the details. This is the first step to make `CUDAGraphRunner` more modularized. We will address the issues in future PRs.

**Review by yuxianq — APPROVED** [review]

**Review by lfr-0531 — APPROVED** [review]

**Comment by QiJune** [comment]
/bot skip --comment "The single test failure is a known issue that was waived in PR https://github.com/NVIDIA/TensorRT-LLM/pull/7194"
